src/transform/src/semijoin_idempotence.rs

```diff
@@ -222,11 +228,11 @@ fn attempt_join_simplification(
                         // The pushdown is for the benefit of CSE on the `A` expressions,
                         // in the not uncommon case of nullable foreign keys in outer joins.
                         // TODO: Discover the transform that would not require this code.
-                        let typ0 = inputs[0].typ().column_types;
-                        let typ1 = inputs[1].typ().column_types;
                         let mut is_not_nulls = Vec::new();
                         for (col0, col1) in ltr.iter() {
-                            if !typ1[*col1].nullable && typ0[*col0].nullable {
+                            if !typ1.column_types[*col1].nullable
+                                && typ0.column_types[*col0].nullable
+                            {
```

Comment:
In theory, when doing the `inputs[1] = candidate.replacement.project(permutation);` above, the types might change (e.g., whether we can determine nullability). But this does not seem to be happening in practice: I added an assert comparing the saved type with a fresh `typ()` call, and ran a bunch of tests (including RQG `left-join-stacks`), and the assert never fired. So this seems fine!

Reply:
Oh, interesting point! I left a comment just so we don't forget about this possible source of inconsistency.